File: app/src/main/java/com/baidu/tv/player/geocoding/AndroidGeocoderStrategy.java
package com.baidu.tv.player.geocoding;

import android.content.Context;
import android.location.Address;
import android.location.Geocoder;
import android.util.Log;

import java.util.List;
import java.util.Locale;

/**
 * Android原生Geocoder策略
 * 优点：无需API Key、系统内置、支持离线
 * 缺点：在中国境内依赖Google服务、可能不可用
 */
public class AndroidGeocoderStrategy implements GeocodingStrategy {
    
    private static final String TAG = "AndroidGeocoderStrategy";
    private static final int TIMEOUT = 5000; // 5秒
    
    @Override
    public String getName() {
        return "AndroidGeocoder";
    }
    
    @Override
    public boolean isAvailable(Context context) {
        return Geocoder.isPresent();
    }
    
    @Override
    public int getPriority() {
        return 2; // 中等优先级
    }
    
    @Override
    public String getAddress(Context context, double latitude, double longitude) {
        try {
            Log.d(TAG, "开始Android Geocoder逆地理编码");
            
            Geocoder geocoder = new Geocoder(context, Locale.CHINESE);
            
            // 获取多个结果以找到最详细的地址
            List<Address> addresses = geocoder.getFromLocation(latitude, longitude, 3);
            
            if (addresses != null && !addresses.isEmpty()) {
                Log.d(TAG, "Geocoder返回了 " + addresses.size() + " 个结果");
                
                // 遍历所有结果，找到最详细的地址
                for (int i = 0; i < addresses.size(); i++) {
                    Address address = addresses.get(i);
                    
                    // 调试：打印地址信息
                    Log.d(TAG, String.format("地址[%d]: %s, %s, %s", 
                            i, address.getFeatureName(), address.getThoroughfare(), address.getLocality()));
                    
                    StringBuilder sb = new StringBuilder();
                    
                    // 优先使用 featureName (建筑物/地标名称)
                    if (address.getFeatureName() != null && !address.getFeatureName().isEmpty()) {
                        sb.append(address.getFeatureName());
                    }
                    
                    // 添加 subThoroughfare (门牌号)
                    if (address.getSubThoroughfare() != null) {
                        if (sb.length() > 0) {
                            sb.append(" ");
                        }
                        sb.append(address.getSubThoroughfare());
                    }
                    
                    // 添加 thoroughfare (街道名称)
                    if (address.getThoroughfare() != null) {
                        if (sb.length() > 0) {
                            sb.append(" ");
                        }
                        sb.append(address.getThoroughfare());
                    }
                    
                    // 添加 subLocality (社区/街道办)
                    if (address.getSubLocality() != null) {
                        if (sb.length() > 0) {
                            sb.append(", ");
                        }
                        sb.append(address.getSubLocality());
                    }
                    
                    // 添加 subAdminArea (区/县)
                    if (address.getSubAdminArea() != null) {
                        if (sb.length() > 0) {
                            sb.append(", ");
                        }
                        sb.append(address.getSubAdminArea());
                    }
                    
                    // 添加 locality (城市)
                    if (address.getLocality() != null) {
                        if (sb.length() > 0) {
                            sb.append(", ");
                        }
                        sb.append(address.getLocality());
                    }
                    
                    // 添加 adminArea (省/州)
                    if (address.getAdminArea() != null) {
                        if (sb.length() > 0) {
                            sb.append(", ");
                        }
                        sb.append(address.getAdminArea());
                    }
                    
                    String detailedResult = sb.toString();
                    // 如果找到了包含街道或建筑物的详细地址，直接返回
                    if (!detailedResult.isEmpty() && 
                        (address.getThoroughfare() != null || address.getFeatureName() != null)) {
                        Log.d(TAG, "✅ Geocoder详细地址: " + detailedResult);
                        return detailedResult;
                    }
                }
                
                // 如果没有找到详细地址，使用第一个结果
                Address address = addresses.get(0);
                StringBuilder sb = new StringBuilder();
                if (address.getSubAdminArea() != null) sb.append(address.getSubAdminArea());
                if (address.getLocality() != null) {
                    if (sb.length() > 0) sb.append(", ");
                    sb.append(address.getLocality());
                }
                if (address.getAdminArea() != null) {
                    if (sb.length() > 0) sb.append(", ");
                    sb.append(address.getAdminArea());
                }
                String basicResult = sb.toString();
                if (!basicResult.isEmpty()) {
                    Log.d(TAG, "✅ Geocoder基础地址: " + basicResult);
                    return basicResult;
                }
            } else {
                Log.d(TAG, "⚠️ Geocoder未返回任何结果");
            }
            
        } catch (Exception e) {
            Log.e(TAG, "❌ Geocoder失败: " + e.getMessage());
        }
        
        return null;
    }
    
    @Override
    public int getTimeout() {
        return TIMEOUT;
    }
    
    @Override
    public String getDescription() {
        return "Android原生Geocoder服务（依赖Google服务）";
    }
}

File: app/src/main/java/com/baidu/tv/player/kt/location/geocoding/AndroidGeocoderStrategy.kt
package com.baidu.tv.player.kt.location.geocoding

import android.content.Context
import android.location.Geocoder
import com.baidu.tv.player.kt.location.GpsCoordinate
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runInterruptible
import java.util.Locale
import javax.inject.Inject

/**
 * Android 系统内置 Geocoder 逆地理编码策略（优先级中）。
 *
 * - 仅当 [Geocoder.isPresent] 为 true 时可用（依赖设备是否内置后端）；
 * - 使用 WGS84 坐标直接查询（系统 Geocoder 按 WGS84）；
 * - 失败/无结果静默返回 null。
 *
 * 注意：此策略不请求任何定位权限，仅将传入坐标反解为地址。
 */
class AndroidGeocoderStrategy @Inject constructor(
    @ApplicationContext private val context: Context,
) : GeocodingStrategy {

    override val name: String = "AndroidGeocoder"
    override val priority: Int = PRIORITY

    override fun isAvailable(): Boolean = Geocoder.isPresent()

    @Suppress("DEPRECATION")
    override suspend fun getAddress(coordinate: GpsCoordinate): String? {
        if (!isAvailable()) return null
        return try {
            runInterruptible(Dispatchers.IO) {
                val geocoder = Geocoder(context, Locale.CHINESE)
                val results = geocoder.getFromLocation(coordinate.latitude, coordinate.longitude, 1)
                val address = results?.firstOrNull() ?: return@runInterruptible null
                buildString {
                    address.adminArea?.let { append(it) }
                    address.locality?.takeIf { it != address.adminArea }?.let { append(it) }
                    address.subLocality?.let { append(it) }
                    address.thoroughfare?.let { append(it) }
                    address.featureName?.takeIf { it != address.thoroughfare }?.let { append(it) }
                }.takeIf { it.isNotBlank() } ?: address.getAddressLine(0)
            }
        } catch (e: CancellationException) {
            throw e
        } catch (_: Exception) {
            null
        }
    }

    companion object {
        const val PRIORITY = 2
    }
}
